opm/autodiff/BlackoilOutputEbos.hpp

```diff
@@ -0,0 +1,391 @@
+    #if HAVE_PTHREAD
+                const bool asyncOutputDefault = true;
+    #else
+                const bool asyncOutputDefault = false;
+    #endif
+                if( param.getDefault("async_output", asyncOutputDefault ) )
+                {
+                    const bool isIORank = parallelOutput_ ? parallelOutput_->isIORank() : true;
+    #if HAVE_PTHREAD
+                    asyncOutput_.reset( new ThreadHandle( isIORank ) );
+    #else
+                    OPM_THROW(std::runtime_error,"Pthreads were not found, cannot enable async_output");
+    #endif
+                }
+            }
+        }
+
+
+
+
+        /*!
+         * \brief Write a blackoil reservoir state to disk for later inspection with
+         *        visualization tools like ResInsight. This function will extract the
+         *        requested output cell properties specified by the RPTRST keyword
+         *        and write these to file.
+         */
+        template<class SimulationDataContainer, class Model>
+        void writeTimeStep(const SimulatorTimerInterface& timer,
+                           const SimulationDataContainer& reservoirStateDummy,
+                           const Opm::WellStateFullyImplicitBlackoil& /*wellStateDummy*/,
+                           const Model& physicalModel,
+                           const bool substep = false,
+                           const double nextstep = -1.0,
+                           const SimulatorReport& simulatorReport = SimulatorReport())
+        {
+            data::Solution fip{};
+
+            if( output_ )
+            {
+                // Get FIP dat
+                getSummaryData( fip, phaseUsage_, physicalModel, summaryConfig() );
+
+                // Add TCPU if simulatorReport is not defaulted.
+                const double totalSolverTime = simulatorReport.solver_time;
+
+                const Opm::WellStateFullyImplicitBlackoil& localWellState = physicalModel.wellModel().wellState();
+
+                if( parallelOutput_ && parallelOutput_->isParallel() )
+                {
+                    // If this is not the initial write and no substep, then the well
+                    // state used in the computation is actually the one of the last
+                    // step. We need that well state for the gathering. Otherwise
+                    // It an exception with a message like "global state does not
+                    // contain well ..." might be thrown.
+                    // The distribution of data::solution is not done here
+                    data::Solution localCellDataDummy{};
+                    int wellStateStepNumber = ( ! substep && timer.reportStepNum() > 0) ?
+                                (timer.reportStepNum() - 1) : timer.reportStepNum();
+                    // collect all solutions to I/O rank
+                    parallelOutput_->collectToIORank( reservoirStateDummy, localWellState,
```

Comment:
Please comment on why this is needed here and what exactly it does (as there are a lot of dummies).
I did look at OPM/ewoms#262 and it seems to me that the functionality for gathering the data on the master rank is replicated there and actually used to collect data.  This is triggered in line 171 `ebosSimulator_.problem().writeOutput` and actually done in the EclipseWriter of ewoms. That means that all the maps are there twice and we build them twice. Maybe this is the cause for the performance regression? 

Maybe it is possible to setup and use the data for gathering the data on master only once?

Reply:
> Please comment on why this is needed here and what exactly it does (as there are a lot of dummies).

From the PR message: 
"The BlackoilOutputEbos is a stripped down version of SimulatorFullyImplicitBlackoilOutput.hpp
that handles the wells and FIP. My plan is to move this functionality to ebos and remove BlackoilOutputEbos when everything is passed."

The reason for the dummies is to be compatible with the legacy code. 

> Maybe this is the cause for the performance regression?

The maps are only made initially (If I have understood the code correctly), It shouldn't be too expensive. But maybe I am wrong.

Reply:
The reason for the regression was the fact that the ebos code computed some extra output values that was not asked for. Now the output is 3 times faster for both norne and model 2.

Reply:
Sounds good. I assume that is mostly due to less copying of data?

Reply:
Yes.  In addition I only supply the output writer with auxiliary values like density, viscosity, bobble point pressures  etc. on reports steps when it is supposed to output. For model 2 output is only every 6 mnd.